File: external/internal/protocol/gateway.go
package protocol

// Gateway framing.
//
// The Central Server opens one outbound WebSocket per World and presents its
// Gateway Credential in the Authorization header. Frames afterwards rely on WSS
// plus the authenticated Gateway Session rather than a second message HMAC, so
// this layer validates structure, version, limits, and correlation only.

var gatewayHelloShape = shape{
	required: map[string]string{
		"v": "integer", "world_id": "id", "central_id": "id",
		"last_topology_revision": "revision", "last_traffic_sequence": "non_negative",
	},
}

var gatewayWelcomeShape = shape{
	required: map[string]string{
		"v": "integer", "gateway_session_id": "id",
		"accepted_topology_revision": "revision", "accepted_traffic_sequence": "non_negative",
		"server_time": "timestamp",
	},
}

var gatewayFrameShape = shape{
	required: map[string]string{"v": "integer", "kind": "operation_name", "body": "object"},
	optional: map[string]string{"request_id": "id", "command_id": "id"},
}

// GatewayHello is the Central Server's opening frame.
type GatewayHello struct {
	WorldID              string
	CentralID            string
	LastTopologyRevision int64
	LastTrafficSequence  int64
}

// Object renders the hello frame.
func (h GatewayHello) Object() Object {
	return Object{
		"v": Version, "world_id": h.WorldID, "central_id": h.CentralID,
		"last_topology_revision": h.LastTopologyRevision,
		"last_traffic_sequence":  h.LastTrafficSequence,
	}
}

// GatewayWelcome is the External Application's reply that opens a Gateway
// Session for one World.
type GatewayWelcome struct {
	GatewaySessionID         string
	AcceptedTopologyRevision int64
	AcceptedTrafficSequence  int64
	ServerTime               string
}

// Object renders the welcome frame.
func (w GatewayWelcome) Object() Object {
	return Object{
		"v": Version, "gateway_session_id": w.GatewaySessionID,
		"accepted_topology_revision": w.AcceptedTopologyRevision,
		"accepted_traffic_sequence":  w.AcceptedTrafficSequence,
		"server_time":                w.ServerTime,
	}
}

// GatewayFrame is one validated Gateway message.
type GatewayFrame struct {
	Kind      string
	RequestID string
	CommandID string
	Body      Object
}

func decodeGatewayObject(text string) (Object, error) {
	if len(text) > GatewayFrameBytes {
		return nil, newError(CodeMessageTooLarge, "gateway message exceeds %d bytes", GatewayFrameBytes)
	}
	value, err := Decode(text, DefaultLimits())
	if err != nil {
		return nil, err
	}
	object, ok := value.(Object)
	if !ok {
		return nil, newError(CodeInvalidMessage, "gateway message must be an object")
	}
	return object, nil
}

func checkGatewayVersion(object Object, what string) error {
	if version, _ := integerValue(object["v"]); version != Version {
		return newError(CodeUnsupportedVersion, "%s declares version %v", what, object["v"])
	}
	return nil
}

// EncodeGatewayHello renders the opening frame.
func EncodeGatewayHello(hello GatewayHello) (string, error) {
	object := hello.Object()
	if err := validateShape(gatewayHelloShape, object, "gateway_hello"); err != nil {
		return "", err
	}
	return Encode(object)
}

// DecodeGatewayHello validates the opening frame.
func DecodeGatewayHello(text string) (*GatewayHello, error) {
	object, err := decodeGatewayObject(text)
	if err != nil {
		return nil, err
	}
	if err := validateShape(gatewayHelloShape, object, "gateway_hello"); err != nil {
		return nil, err
	}
	if err := checkGatewayVersion(object, "gateway_hello"); err != nil {
		return nil, err
	}
	worldID, _ := stringValue(object["world_id"])
	centralID, _ := stringValue(object["central_id"])
	topology, _ := integerValue(object["last_topology_revision"])
	traffic, _ := integerValue(object["last_traffic_sequence"])
	return &GatewayHello{
		WorldID: worldID, CentralID: centralID,
		LastTopologyRevision: topology, LastTrafficSequence: traffic,
	}, nil
}

// EncodeGatewayWelcome renders the accepting reply.
func EncodeGatewayWelcome(welcome GatewayWelcome) (string, error) {
	object := welcome.Object()
	if err := validateShape(gatewayWelcomeShape, object, "gateway_welcome"); err != nil {
		return "", err
	}
	return Encode(object)
}

// DecodeGatewayWelcome validates the accepting reply.
func DecodeGatewayWelcome(text string) (*GatewayWelcome, error) {
	object, err := decodeGatewayObject(text)
	if err != nil {
		return nil, err
	}
	if err := validateShape(gatewayWelcomeShape, object, "gateway_welcome"); err != nil {
		return nil, err
	}
	if err := checkGatewayVersion(object, "gateway_welcome"); err != nil {
		return nil, err
	}
	sessionID, _ := stringValue(object["gateway_session_id"])
	topology, _ := integerValue(object["accepted_topology_revision"])
	traffic, _ := integerValue(object["accepted_traffic_sequence"])
	serverTime, _ := stringValue(object["server_time"])
	return &GatewayWelcome{
		GatewaySessionID: sessionID, AcceptedTopologyRevision: topology,
		AcceptedTrafficSequence: traffic, ServerTime: serverTime,
	}, nil
}

// EncodeGatewayFrame renders one Gateway message.
func EncodeGatewayFrame(frame GatewayFrame) (string, error) {
	if !Allows(TransportGateway, frame.Kind) {
		return "", newError(CodeInvalidMessage, "kind %q is not carried by the Gateway", frame.Kind)
	}
	if err := ValidateBody(frame.Kind, frame.Body); err != nil {
		return "", err
	}
	object := Object{"v": Version, "kind": frame.Kind, "body": frame.Body}
	if frame.RequestID != "" {
		object["request_id"] = frame.RequestID
	}
	if frame.CommandID != "" {
		object["command_id"] = frame.CommandID
	}
	if err := validateShape(gatewayFrameShape, object, "gateway_frame"); err != nil {
		return "", err
	}
	text, err := Encode(object)
	if err != nil {
		return "", err
	}
	if err := checkGatewayFrameSize(frame.Kind, text); err != nil {
		return "", err
	}
	return text, nil
}

// DecodeGatewayFrame validates one Gateway message.
func DecodeGatewayFrame(text string) (*GatewayFrame, error) {
	object, err := decodeGatewayObject(text)
	if err != nil {
		return nil, err
	}
	if err := validateShape(gatewayFrameShape, object, "gateway_frame"); err != nil {
		return nil, err
	}
	if err := checkGatewayVersion(object, "gateway_frame"); err != nil {
		return nil, err
	}
	kind, _ := stringValue(object["kind"])
	if !Allows(TransportGateway, kind) {
		return nil, newError(CodeInvalidMessage, "kind %q is not carried by the Gateway", kind)
	}
	if err := checkGatewayFrameSize(kind, text); err != nil {
		return nil, err
	}
	body, _ := object["body"].(Object)
	if err := ValidateBody(kind, body); err != nil {
		return nil, err
	}
	requestID, _ := stringValue(object["request_id"])
	commandID, _ := stringValue(object["command_id"])
	return &GatewayFrame{Kind: kind, RequestID: requestID, CommandID: commandID, Body: body}, nil
}

// checkGatewayFrameSize applies the narrower batch ceiling on top of the
// Gateway message limit. CraftNet performs no fragmentation: an oversized batch
// must be split by its sender.
func checkGatewayFrameSize(kind, text string) error {
	if kind == "traffic_batch" && len(text) > TrafficBatchBytes {
		return newError(CodeMessageTooLarge, "traffic batch exceeds %d bytes", TrafficBatchBytes)
	}
	if len(text) > GatewayFrameBytes {
		return newError(CodeMessageTooLarge, "gateway message exceeds %d bytes", GatewayFrameBytes)
	}
	return nil
}

File: packages/craftnet-protocol/files/gateway.lua
-- Gateway framing.
--
-- The Central Server opens one outbound WebSocket per World and presents its
-- Gateway Credential in the Authorization header. Frames afterwards rely on WSS
-- plus the authenticated Gateway Session rather than a second message HMAC, so
-- this layer validates structure, version, limits, and correlation only -- and
-- nothing here signs, derives, or holds a secret.
--
-- It mirrors external/internal/protocol/gateway.go field for field. The two
-- ends of one socket disagreeing about the envelope is exactly the class of
-- mistake the shared fixture catalog exists to catch.

local internal = ...
local cj1 = internal("cj1")
local schema = internal("schema")
local limits = internal("limits")

local gateway = {}

local helloShape = {
  required = {
    v = "integer", world_id = "id", central_id = "id",
    last_topology_revision = "revision", last_traffic_sequence = "non_negative",
  },
}

local welcomeShape = {
  required = {
    v = "integer", gateway_session_id = "id",
    accepted_topology_revision = "revision", accepted_traffic_sequence = "non_negative",
    server_time = "timestamp",
  },
}

local frameShape = {
  required = { v = "integer", kind = "operation_name", body = "object" },
  optional = { request_id = "id", command_id = "id" },
}

-- checkSize applies the narrower batch ceiling on top of the Gateway message
-- limit. CraftNet performs no fragmentation: an oversized batch is split by
-- whoever built it, never here.
local function checkSize(kind, text)
  if kind == "traffic_batch" and #text > limits.TRAFFIC_BATCH_BYTES then
    return false, "traffic batch exceeds " .. limits.TRAFFIC_BATCH_BYTES .. " bytes",
      "message_too_large"
  end
  if #text > limits.GATEWAY_FRAME_BYTES then
    return false, "gateway message exceeds " .. limits.GATEWAY_FRAME_BYTES .. " bytes",
      "message_too_large"
  end
  return true
end

local function decodeObject(text)
  if type(text) ~= "string" then
    return nil, "invalid_message", "a gateway message must be text"
  end
  if #text > limits.GATEWAY_FRAME_BYTES then
    return nil, "message_too_large",
      "gateway message exceeds " .. limits.GATEWAY_FRAME_BYTES .. " bytes"
  end
  local value, code, problem = cj1.decode(text)
  if not value then return nil, code or "invalid_message", problem end
  if not schema.scalars.object(value) then
    return nil, "invalid_message", "a gateway message must be an object"
  end
  return value
end

local function checkVersion(object, what)
  if rawget(object, "v") ~= schema.VERSION then
    return false, "unsupported_version",
      what .. " declares version " .. tostring(rawget(object, "v"))
  end
  return true
end

local function validated(shape, object, what)
  local ok, problem, code = schema.validateShape(shape, object, what)
  if not ok then return nil, code or "invalid_message", problem end
  return true
end

--------------------------------------------------------------------------
-- The opening exchange
--------------------------------------------------------------------------

-- encodeHello renders what the Central Server sends first. The revisions it
-- names are what the External Application answers against, so a reconnecting
-- World is told what it already has rather than resending everything.
function gateway.encodeHello(fields)
  local object = cj1.object({
    v = schema.VERSION,
    world_id = fields.world_id,
    central_id = fields.central_id,
    last_topology_revision = fields.last_topology_revision or 0,
    last_traffic_sequence = fields.last_traffic_sequence or 0,
  })
  local ok, code, problem = validated(helloShape, object, "gateway_hello")
  if not ok then return nil, code, problem end
  local text, encodeCode, encodeProblem = cj1.encode(object)
  if not text then return nil, encodeCode or "invalid_message", encodeProblem end
  return text
end

-- decodeWelcome reads the reply that opens a Gateway Session. Until it decodes,
-- there is no session: a Central Server that cannot read the welcome has not
-- connected, however healthy the socket looks.
function gateway.decodeWelcome(text)
  local object, code, problem = decodeObject(text)
  if not object then return nil, code, problem end
  local ok, shapeCode, shapeProblem = validated(welcomeShape, object, "gateway_welcome")
  if not ok then return nil, shapeCode, shapeProblem end
  local sound, versionCode, versionProblem = checkVersion(object, "gateway_welcome")
  if not sound then return nil, versionCode, versionProblem end
  return {
    gateway_session_id = rawget(object, "gateway_session_id"),
    accepted_topology_revision = rawget(object, "accepted_topology_revision"),
    accepted_traffic_sequence = rawget(object, "accepted_traffic_sequence"),
    server_time = rawget(object, "server_time"),
  }
end

--------------------------------------------------------------------------
-- Frames
--------------------------------------------------------------------------

-- encodeFrame renders one Gateway message. A kind the Gateway does not carry is
-- refused here rather than sent and refused at the far end.
function gateway.encodeFrame(kind, body, correlation)
  correlation = correlation or {}
  if not schema.allows("gateway", kind) then
    return nil, "invalid_message", "'" .. tostring(kind) .. "' is not carried by the Gateway"
  end
  local ok, problem, code = schema.validateBody(kind, body)
  if not ok then return nil, code or "invalid_message", problem end

  local object = cj1.object({ v = schema.VERSION, kind = kind, body = body })
  if correlation.request_id then rawset(object, "request_id", correlation.request_id) end
  if correlation.command_id then rawset(object, "command_id", correlation.command_id) end

  local shaped, shapeCode, shapeProblem = validated(frameShape, object, "gateway_frame")
  if not shaped then return nil, shapeCode, shapeProblem end

  local text, encodeCode, encodeProblem = cj1.encode(object)
  if not text then return nil, encodeCode or "invalid_message", encodeProblem end
  local sized, sizeProblem, sizeCode = checkSize(kind, text)
  if not sized then return nil, sizeCode, sizeProblem end
  return text
end

-- decodeFrame validates one Gateway message. Everything a caller then acts on
-- has already been checked: the version, the envelope, the kind against the
-- transport, the size, and the body against its schema.
function gateway.decodeFrame(text)
  local object, code, problem = decodeObject(text)
  if not object then return nil, code, problem end
  local ok, shapeCode, shapeProblem = validated(frameShape, object, "gateway_frame")
  if not ok then return nil, shapeCode, shapeProblem end
  local sound, versionCode, versionProblem = checkVersion(object, "gateway_frame")
  if not sound then return nil, versionCode, versionProblem end

  local kind = rawget(object, "kind")
  if not schema.allows("gateway", kind) then
    return nil, "invalid_message", "'" .. tostring(kind) .. "' is not carried by the Gateway"
  end
  local sized, sizeProblem, sizeCode = checkSize(kind, text)
  if not sized then return nil, sizeCode, sizeProblem end

  local body = rawget(object, "body")
  local valid, bodyProblem, bodyCode = schema.validateBody(kind, body)
  if not valid then return nil, bodyCode or "invalid_message", bodyProblem end

  return {
    kind = kind,
    request_id = rawget(object, "request_id"),
    command_id = rawget(object, "command_id"),
    body = body,
  }
end

return gateway
